File: internal/ebpf/programs/c/tc_offload.c
#include "vmlinux.h"
#include "bpf_helpers.h"
#include "bpf_endian.h"
#include "common.h"

// Flow verdicts
#define VERDICT_UNKNOWN  0
#define VERDICT_TRUSTED  1
#define VERDICT_DROP     2

// nftables marks for flow bypass
#define NFQUEUE_BYPASS_MARK 0x200000  // Skip NFQUEUE, go to acceptance

// QoS profile structure
struct qos_profile {
    __u32 rate_limit;
    __u32 burst_limit;
    __u8 priority;
    __u8 app_class;
    __u8 padding[2];
};

// Maps
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __type(key, struct flow_key);
    __type(value, struct flow_state);
    __uint(max_entries, 100000);
    __uint(pinning, LIBBPF_PIN_BY_NAME);
} flow_map SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __type(key, __u32);
    __type(value, struct qos_profile);
    __uint(max_entries, 16);
    __uint(pinning, LIBBPF_PIN_BY_NAME);
} qos_profiles SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
    __type(key, __u32);
    __type(value, struct tc_stats);
    __uint(max_entries, 1);
} tc_stats_map SEC(".maps");

// TC statistics
struct tc_stats {
	__u64 packets_processed;
	__u64 packets_fast_path;
	__u64 packets_slow_path;
	__u64 packets_dropped;
	__u64 bytes_processed;
};

// Helper functions
static __always_inline int extract_flow_key(struct __sk_buff *skb, struct flow_key *key)
{
	void *data_end = (void *)(long)skb->data_end;
	void *data = (void *)(long)skb->data;

	// Minimum packet size for Ethernet + IP
	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
		return -1;

	struct ethhdr *eth = data;
	struct iphdr *ip = (struct iphdr *)(eth + 1);

	// Only handle IPv4 for now
	if (eth->h_proto != __bpf_constant_htons(ETH_P_IP))
		return -1;

	// Extract basic 5-tuple
	key->src_ip = ip->saddr;
	key->dst_ip = ip->daddr;
	key->ip_proto = ip->protocol;

	// Handle transport layer ports
	if (ip->protocol == IPPROTO_TCP) {
		struct tcphdr *tcp = (struct tcphdr *)((void *)ip + (ip->ihl * 4));
		if ((void *)tcp + sizeof(struct tcphdr) > data_end)
			return -1;

		key->src_port = __bpf_ntohs(tcp->source);
		key->dst_port = __bpf_ntohs(tcp->dest);
	} else if (ip->protocol == IPPROTO_UDP) {
		struct udphdr *udp = (struct udphdr *)((void *)ip + (ip->ihl * 4));
		if ((void *)udp + sizeof(struct udphdr) > data_end)
			return -1;

		key->src_port = __bpf_ntohs(udp->source);
		key->dst_port = __bpf_ntohs(udp->dest);
	} else {
		key->src_port = 0;
		key->dst_port = 0;
	}

	// Use ingress interface as part of key
	key->ifindex = skb->ifindex;

	return 0;
}

// Apply QoS marking based on flow state
static inline int apply_qos(struct __sk_buff *skb, struct flow_state *state)
{
    if (state->qos_profile == QOS_PROFILE_DEFAULT)
        return TC_ACT_OK;

    struct qos_profile *qos = bpf_map_lookup_elem(&qos_profiles, &state->qos_profile);
    if (!qos)
        return TC_ACT_OK;

    // Set priority for hardware scheduling
    skb->priority = qos->priority;

    // Set queue mapping if needed
    if (qos->app_class == QOS_PROFILE_VIDEO || qos->app_class == QOS_PROFILE_VOICE) {
        skb->queue_mapping = qos->app_class;
    }

    // Mark packet for QoS handling
    skb->mark |= 0x100000;  // QoS mark bit

    return TC_ACT_OK;
}

static __always_inline void update_stats(__u64 packets, __u64 bytes, __u64 fast_path, __u64 slow_path, __u64 dropped)
{
	__u32 key = 0;
	struct tc_stats *stats = bpf_map_lookup_elem(&tc_stats_map, &key);
	if (!stats)
		return;

	__sync_fetch_and_add(&stats->packets_processed, packets);
	__sync_fetch_and_add(&stats->bytes_processed, bytes);
	__sync_fetch_and_add(&stats->packets_fast_path, fast_path);
	__sync_fetch_and_add(&stats->packets_slow_path, slow_path);
	__sync_fetch_and_add(&stats->packets_dropped, dropped);
}

// Main TC classifier
SEC("tc")
int tc_fast_path(struct __sk_buff *skb)
{
	struct flow_key key = {};

	// Extract flow key
	if (extract_flow_key(skb, &key) < 0) {
		// Can't extract flow info, let it pass to normal processing
		update_stats(1, skb->len, 0, 1, 0);
		return TC_ACT_OK;
	}

	// Lookup flow state in the map
	struct flow_state *state = bpf_map_lookup_elem(&flow_map, &key);

	if (!state) {
		// Unknown flow -> Pass to Stack -> NFQUEUE
		// This is the slow path for new flows
		update_stats(1, skb->len, 0, 1, 0);
		return TC_ACT_OK;
	}

	// Update flow statistics
	__sync_fetch_and_add(&state->packet_count, 1);
	__sync_fetch_and_add(&state->byte_count, skb->len);
	__sync_fetch_and_add(&state->last_seen, bpf_ktime_get_ns());

	// Check flow verdict
	if (state->verdict == VERDICT_TRUSTED) {
		// Trusted flow - mark for nftables bypass
		skb->mark = NFQUEUE_BYPASS_MARK;

		// Apply QoS marking if configured
		int qos_result = apply_qos(skb, state);
		if (qos_result != TC_ACT_OK) {
			update_stats(1, skb->len, 0, 0, 1);
			return qos_result;
		}

		// Update fast path stats
		update_stats(1, skb->len, 1, 0, 0);

		// Optional: If we have a simple redirect case, we could use:
		// return bpf_redirect(skb->ifindex, 0);

		return TC_ACT_OK;
	} else if (state->verdict == VERDICT_DROP) {
		// Blocked flow - drop early
		update_stats(1, skb->len, 0, 0, 1);
		return TC_ACT_SHOT;
	}

	// Unknown verdict - pass to normal processing
	update_stats(1, skb->len, 0, 1, 0);
	return TC_ACT_OK;
}

// TC egress program for outbound traffic
SEC("tc")
int tc_egress_fast_path(struct __sk_buff *skb)
{
	struct flow_key key = {};

	// Extract flow key (reverse direction for egress)
	if (extract_flow_key(skb, &key) < 0) {
		return TC_ACT_OK;
	}

	// Swap src/dst for egress lookup
	__u32 tmp_ip = key.src_ip;
	key.src_ip = key.dst_ip;
	key.dst_ip = tmp_ip;

	__u16 tmp_port = key.src_port;
	key.src_port = key.dst_port;
	key.dst_port = tmp_port;

	// Lookup flow state
	struct flow_state *state = bpf_map_lookup_elem(&flow_map, &key);

	if (!state) {
		return TC_ACT_OK;
	}

	// Update flow statistics
	__sync_fetch_and_add(&state->packet_count, 1);
	__sync_fetch_and_add(&state->byte_count, skb->len);

	// Apply verdict
	if (state->verdict == VERDICT_TRUSTED) {
		skb->mark = NFQUEUE_BYPASS_MARK;
		return TC_ACT_OK;
	} else if (state->verdict == VERDICT_DROP) {
		return TC_ACT_SHOT;
	}

	return TC_ACT_OK;
}

File: internal/ebpf/programs/tc_offload.go
// Copyright (C) 2026 Ben Grimm. Licensed under AGPL-3.0 (https://www.gnu.org/licenses/agpl-3.0.txt)

package programs

import (
	"fmt"
	"net"
	"runtime"
	"time"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"

	"grimm.is/flywall/internal/ebpf/types"
	"grimm.is/flywall/internal/logging"
)

// TCOffloadProgram manages the TC fast-path program
type TCOffloadProgram struct {
	collection *ebpf.Collection
	links      []link.Link
	logger     *logging.Logger
	stats      *TCStats
}

// TCStats holds TC program statistics
type TCStats struct {
	PacketsProcessed uint64 `json:"packets_processed"`
	PacketsFastPath  uint64 `json:"packets_fast_path"`
	PacketsSlowPath  uint64 `json:"packets_slow_path"`
	PacketsDropped   uint64 `json:"packets_dropped"`
	BytesProcessed   uint64 `json:"bytes_processed"`
}

// NewTCOffloadProgram loads and attaches the TC offload program
func NewTCOffloadProgram(logger *logging.Logger) (*TCOffloadProgram, error) {
	// Load pre-compiled program from embedded bytecode
	spec, err := LoadTcOffload()
	if err != nil {
		return nil, fmt.Errorf("failed to load TC offload spec: %w", err)
	}

	// Disable pinning for all maps (not needed in most environments/tests)
	for _, m := range spec.Maps {
		m.Pinning = ebpf.PinNone
	}

	// Workaround for tc_stats_map issues on some kernels/environments
	if m, ok := spec.Maps["tc_stats_map"]; ok {
		m.Type = ebpf.Array
	}

	// Increase map sizes for production
	if spec.Maps["flow_map"] != nil {
		spec.Maps["flow_map"].MaxEntries = 100000
	}

	// Load the collection
	collection, err := ebpf.NewCollection(spec)
	if err != nil {
		return nil, fmt.Errorf("failed to load TC offload collection: %w", err)
	}

	program := &TCOffloadProgram{
		collection: collection,
		links:      make([]link.Link, 0),
		logger:     logger,
		stats:      &TCStats{},
	}

	// Initialize statistics map
	if err := program.initStats(); err != nil {
		program.Close()
		return nil, fmt.Errorf("failed to initialize stats: %w", err)
	}

	return program, nil
}

// Attach attaches the TC program to the specified interface
func (p *TCOffloadProgram) Attach(ifaceName string) error {
	iface, err := net.InterfaceByName(ifaceName)
	if err != nil {
		return fmt.Errorf("failed to find interface %s: %w", ifaceName, err)
	}

	// Attach ingress TC program
	ingressProg := p.collection.Programs["tc_fast_path"]
	if ingressProg == nil {
		return fmt.Errorf("tc_fast_path program not found")
	}

	// For now, skip actual attachment on macOS
	if runtime.GOOS != "linux" {
		p.logger.Info("Skipping TC attachment on non-Linux platform", "interface", ifaceName)
		return nil
	}

	ingressLink, err := link.AttachTCX(link.TCXOptions{
		Program:   ingressProg,
		Interface: iface.Index,
		Attach:    ebpf.AttachTCXIngress,
	})
	if err != nil {
		return fmt.Errorf("failed to attach ingress TC program: %w", err)
	}

	p.links = append(p.links, ingressLink)
	p.logger.Info("Attached TC ingress program", "interface", ifaceName)

	// Attach egress TC program
	egressProg := p.collection.Programs["tc_egress_fast_path"]
	if egressProg == nil {
		return fmt.Errorf("tc_egress_fast_path program not found")
	}

	if runtime.GOOS != "linux" {
		p.logger.Info("Skipping TC egress attachment on non-Linux platform", "interface", ifaceName)
		return nil
	}

	egressLink, err := link.AttachTCX(link.TCXOptions{
		Program:   egressProg,
		Interface: iface.Index,
		Attach:    ebpf.AttachTCXEgress,
	})
	if err != nil {
		// Cleanup ingress link on failure
		ingressLink.Close()
		p.links = p.links[:len(p.links)-1]
		return fmt.Errorf("failed to attach egress TC program: %w", err)
	}

	p.links = append(p.links, egressLink)
	p.logger.Info("Attached TC egress program", "interface", ifaceName)

	return nil
}

// Detach removes the TC program from all interfaces
func (p *TCOffloadProgram) Detach() error {
	var lastErr error

	for _, link := range p.links {
		if err := link.Close(); err != nil {
			lastErr = err
		}
	}

	p.links = p.links[:0]

	return lastErr
}

// FlowMap returns the flow state map
func (p *TCOffloadProgram) FlowMap() *ebpf.Map {
	return p.collection.Maps["flow_map"]
}

// StatsMap returns the statistics map
func (p *TCOffloadProgram) StatsMap() *ebpf.Map {
	return p.collection.Maps["tc_stats_map"]
}

// UpdateFlow updates or creates a flow state
func (p *TCOffloadProgram) UpdateFlow(key types.FlowKey, state types.FlowState) error {
	flowMap := p.FlowMap()
	if flowMap == nil {
		return fmt.Errorf("flow map not available")
	}

	// Convert to C structures
	cKey := keyToC(key)
	cState := stateToC(state)

	return flowMap.Update(&cKey, &cState, ebpf.UpdateAny)
}

// DeleteFlow removes a flow state
func (p *TCOffloadProgram) DeleteFlow(key types.FlowKey) error {
	flowMap := p.FlowMap()
	if flowMap == nil {
		return fmt.Errorf("flow map not available")
	}

	cKey := keyToC(key)
	return flowMap.Delete(&cKey)
}

// GetFlow retrieves a flow state
func (p *TCOffloadProgram) GetFlow(key types.FlowKey) (*types.FlowState, error) {
	flowMap := p.FlowMap()
	if flowMap == nil {
		return nil, fmt.Errorf("flow map not available")
	}

	cKey := keyToC(key)
	var cState CFlowState

	if err := flowMap.Lookup(&cKey, &cState); err != nil {
		return nil, err
	}

	state := cStateToState(cState)
	return &state, nil
}

// GetStats returns current TC statistics
func (p *TCOffloadProgram) GetStats() *TCStats {
	statsMap := p.StatsMap()
	if statsMap == nil {
		return p.stats
	}

	var key uint32 = 0
	var cStats CTCStats

	if err := statsMap.Lookup(&key, &cStats); err == nil {
		p.stats.PacketsProcessed = cStats.PacketsProcessed
		p.stats.PacketsFastPath = cStats.PacketsFastPath
		p.stats.PacketsSlowPath = cStats.PacketsSlowPath
		p.stats.PacketsDropped = cStats.PacketsDropped
		p.stats.BytesProcessed = cStats.BytesProcessed
	}

	return p.stats
}

// initStats initializes the statistics map
func (p *TCOffloadProgram) initStats() error {
	statsMap := p.StatsMap()
	if statsMap == nil {
		return fmt.Errorf("stats map not found")
	}

	var key uint32 = 0
	initialStats := CTCStats{}

	return statsMap.Update(&key, &initialStats, ebpf.UpdateAny)
}

// Cleanup removes expired flows
func (p *TCOffloadProgram) Cleanup(timeout time.Duration) error {
	flowMap := p.FlowMap()
	if flowMap == nil {
		return fmt.Errorf("flow map not available")
	}

	now := time.Now().UnixNano()
	timeoutNs := uint64(timeout.Nanoseconds())

	var keys []CFlowKey
	var values []CFlowState

	// Collect all entries
	iterator := flowMap.Iterate()
	for iterator.Next(&keys, &values) {
		if uint64(now)-values[0].LastSeen < timeoutNs {
			// Flow has expired, mark for deletion
			if err := flowMap.Delete(&keys[0]); err != nil {
				p.logger.Error("Failed to delete expired flow", "error", err)
			}
		}
	}

	if err := iterator.Err(); err != nil {
		return fmt.Errorf("failed to iterate flow map: %w", err)
	}

	return nil
}

// Close detaches the program and cleans up resources
func (p *TCOffloadProgram) Close() error {
	if err := p.Detach(); err != nil {
		p.logger.Error("Error detaching TC program", "error", err)
	}

	if p.collection != nil {
		p.collection.Close()
	}

	return nil
}

// GetCollection returns the eBPF collection for statistics
func (p *TCOffloadProgram) GetCollection() *ebpf.Collection {
	return p.collection
}

// C structures for eBPF map compatibility
type CFlowKey struct {
	SrcIP   uint32
	DstIP   uint32
	SrcPort uint16
	DstPort uint16
	IPProto uint8
	Padding [3]uint8
	Ifindex uint32
}

type CFlowState struct {
	Verdict     uint8
	QoSProfile  uint8
	Flags       uint16
	PacketCount uint32
	ByteCount   uint32
	LastSeen    uint64
	CreatedAt   uint64
	ExpiresAt   uint64
	JA3Hash     [4]uint32
	SNI         [64]byte
}

type CTCStats struct {
	PacketsProcessed uint64
	PacketsFastPath  uint64
	PacketsSlowPath  uint64
	PacketsDropped   uint64
	BytesProcessed   uint64
}

type CQoSProfile struct {
	RateLimit  uint32
	BurstLimit uint32
	Priority   uint8
	AppClass   uint8
	Padding    [2]uint8
}

// SetQoSProfile sets the QoS profile for a flow
func (p *TCOffloadProgram) SetQoSProfile(key types.FlowKey, profileID uint8) error {
	flowMap := p.FlowMap()
	if flowMap == nil {
		return fmt.Errorf("flow map not available")
	}

	// Get existing flow state
	state, err := p.GetFlow(key)
	if err != nil {
		return fmt.Errorf("flow not found: %w", err)
	}

	// Update QoS profile
	state.QoSProfile = profileID

	// Update in map
	return p.UpdateFlow(key, *state)
}

// UpdateQoSProfile updates a QoS profile configuration
func (p *TCOffloadProgram) UpdateQoSProfile(profileID uint32, profile types.QoSProfile) error {
	qosMap := p.collection.Maps["qos_profiles"]
	if qosMap == nil {
		return fmt.Errorf("QoS profiles map not available")
	}

	cProfile := CQoSProfile{
		RateLimit:  profile.RateLimit,
		BurstLimit: profile.BurstLimit,
		Priority:   profile.Priority,
		AppClass:   profile.AppClass,
	}

	return qosMap.Update(&profileID, &cProfile, ebpf.UpdateAny)
}

// GetQoSProfile retrieves a QoS profile configuration
func (p *TCOffloadProgram) GetQoSProfile(profileID uint32) (*types.QoSProfile, error) {
	qosMap := p.collection.Maps["qos_profiles"]
	if qosMap == nil {
		return nil, fmt.Errorf("QoS profiles map not available")
	}

	var cProfile CQoSProfile
	if err := qosMap.Lookup(&profileID, &cProfile); err != nil {
		return nil, err
	}

	return &types.QoSProfile{
		RateLimit:  cProfile.RateLimit,
		BurstLimit: cProfile.BurstLimit,
		Priority:   cProfile.Priority,
		AppClass:   cProfile.AppClass,
	}, nil
}

// Helper functions to convert between Go and C structures
func keyToC(key types.FlowKey) CFlowKey {
	return CFlowKey{
		SrcIP:   key.SrcIP,
		DstIP:   key.DstIP,
		SrcPort: key.SrcPort,
		DstPort: key.DstPort,
		IPProto: key.IPProto,
	}
}

func stateToC(state types.FlowState) CFlowState {
	return CFlowState{
		Verdict:     state.Verdict,
		QoSProfile:  state.QoSProfile,
		Flags:       state.Flags,
		PacketCount: uint32(state.PacketCount),
		ByteCount:   uint32(state.ByteCount),
		LastSeen:    state.LastSeen,
		CreatedAt:   state.CreatedAt,
		ExpiresAt:   state.ExpiresAt,
		JA3Hash:     state.JA3Hash,
		SNI:         state.SNI,
	}
}

func cStateToState(state CFlowState) types.FlowState {
	return types.FlowState{
		Verdict:     state.Verdict,
		QoSProfile:  state.QoSProfile,
		Flags:       state.Flags,
		PacketCount: uint64(state.PacketCount),
		ByteCount:   uint64(state.ByteCount),
		LastSeen:    state.LastSeen,
		CreatedAt:   state.CreatedAt,
		ExpiresAt:   state.ExpiresAt,
		JA3Hash:     state.JA3Hash,
		SNI:         state.SNI,
	}
}
